include/pybind11/stl_bind.h

```diff
@@ -129,8 +129,10 @@ void vector_modifiers(enable_if_t<is_copy_constructible<typename Vector::value_t
     }));
 
     cl.def("extend",
-       [](Vector &v, const Vector &src) {
-           v.insert(v.end(), src.begin(), src.end());
+       [](Vector &v, iterable it) {
```

Comment:
Do we really want `iterable`? It doesn't make much sense to extend a vector of strings with a dictionary even if its keys are strings. With this PR `v_int2.extend({1:1,2:2})` would extend the `v_int2` with `[1, 2]`.

Consider strings too, since they are iterable:
```
>>> from foo import VectorString
>>> bar = VectorString()
>>> bar.extend('bar')
>>> bar
VectorString[b,a,r]
```

I personally think that we should only allow python `list` and C++-like `VectorFoo` to extend `VectorFoo`.
Also, there's a performance benefit in not doing the `cast<T>` when `VectorFoo` is passed, so an overload for python types should at least be considered.